syntax = "proto3";

package ocr_service;

service OCRService {
  rpc RecognizeBytes( RecognizeBytesRequest ) returns ( RecognizeDefaultResponse ) {}
  rpc RecognizeBase64( RecognizeBase64Request ) returns ( RecognizeDefaultResponse ) {}
  rpc Detect( DetectRequest ) returns ( DetectResponse ) {}
  rpc GetSupportedLanguages( GetSupportedLanguagesRequest ) returns ( GetSupportedLanguagesResponse ) {}
  rpc GetSupportedModels( GetSupportedModelsRequest ) returns ( GetSupportedModelsResponse ) {}
  rpc InstallModel( InstallModelRequest ) returns ( InstallModelResponse ) {}
  rpc UpdatePpOcrSettings( UpdatePpOcrSettingsRequest ) returns ( UpdateSettingsResponse ) {}
  rpc KeepAlive( KeepAliveRequest ) returns ( KeepAliveResponse ) {}
  rpc MotionDetection( MotionDetectionRequest ) returns ( MotionDetectionResponse ) {}
}

message KeepAliveRequest {
  bool keep_alive = 1;
  int32 timeout_seconds = 2;
}
message KeepAliveResponse {}

message RecognizeBytesRequest {
  string id = 1;
  string language_code = 2;
  bytes image_bytes = 3;
  repeated Box boxes = 4;
  string ocr_engine = 5; // MangaOCR | PaddleOCR
}
message RecognizeBase64Request {
  string id = 1;
  string language_code = 2;
  string base64_image = 3;
  repeated Box boxes = 4;
  string ocr_engine = 5; // MangaOCR | PaddleOCR | AppleVision
}

message Vertex {
  int32 x = 1;
  int32 y = 2;
}
message Box {
  Vertex bottom_left = 1;
  Vertex bottom_right = 2;
  Vertex top_left = 3;
  Vertex top_right = 4;
}
message TextLine {
  string content = 1;
  Box box = 2;
}
message Result {
  repeated TextLine text_lines = 1;
  float recognition_score = 2;
  float classification_score = 3;
  int32 classification_label = 4; // Text direction
  Box box = 5;
  bool is_vertical = 6;
}
message ContextResolution {
  int32 width = 1;
  int32 height = 2;
}
message RecognizeDefaultResponse {
  string id = 1;
  repeated Result results = 2;
  ContextResolution context_resolution = 3;
}


message GetSupportedLanguagesRequest {
  string ocr_engine = 1; // MangaOCR | PaddleOCR | AppleVision
}
message GetSupportedLanguagesResponse {
  repeated string language_codes = 1;
}


message DetectRequest {
  string id = 1;
  string language_code = 2;
  bool crop_image = 3;
  bytes image_bytes = 4;
  string ocr_engine = 5; // MangaOCR | PaddleOCR | AppleVision
}

message DetectionResult {
  bytes image_bytes = 1;
  Box box = 2;
  repeated TextLine text_lines = 3;
}
message DetectResponse {
  string id = 1;
  repeated DetectionResult results = 2;
  ContextResolution context_resolution = 3;
}

message MotionDetectionRequest {
  string stream_id = 1;
  bytes frame = 2;
  int32 threshold_min = 3;
  int32 threshold_max = 4;
  int32 stream_length = 5;
}
message MotionDetectionResponse {
  int32 frame_diff_sum = 1;
  int32 frame_threshold_non_zero_count = 2;
}


message UpdatePpOcrSettingsRequest {
  int32 max_image_width = 1; // ppocr "max_side_length"
  int32 cpu_threads = 2;
  string inference_runtime = 3;
  double det_db_thresh = 4;
  double det_db_box_thresh = 5;
  double det_db_unclip_ratio = 6;
  string det_db_score_mode = 7;
  bool use_dilation = 8;
  double cls_thresh = 9;
}
message UpdateSettingsResponse {
  bool success = 1;
}


message TextRecognitionModel {
  string name = 1;
  repeated string language_codes = 2;
  bool is_installed = 3;
}
message GetSupportedModelsRequest {
  string ocr_engine = 1; // MangaOCR | PaddleOCR | AppleVision
  string language_code = 2;
}
message GetSupportedModelsResponse {
  repeated TextRecognitionModel models = 1;
}

message InstallModelRequest {
  string ocr_engine = 1; // MangaOCR | PaddleOCR | AppleVision
  string model_name = 2;
}
message InstallModelResponse {
  bool success = 1;
}